filebeat/input/v2/statestore/store.go

```diff
@@ -0,0 +1,132 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package statestore
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/registry"
+)
+
+// Store provides some coordinates access to a registry.Store.
+// All update and read operations require users to acquire an resource first.
+// A Resource must be locked before it can be modified. This ensures that at most
+// one go-routine has access to a resource. Lock/TryLock/Unlock can be used to
+// coordinate resource access even between independent components.
+type Store struct {
+	active bool
+	waiter chan struct{}
+
+	log *logp.Logger
+
+	persistentStore *registry.Store
+
+	resourcesMux sync.Mutex
+	resources    table
+}
+
+// NewStore creates a new state Store that is backed by a persistent store.
+func NewStore(log *logp.Logger, store *registry.Store) *Store {
+	invariant(log != nil, "missing a logger")
+	invariant(store != nil, "missing a persistent store")
+
+	return &Store{
+		active:          true,
+		log:             log,
+		persistentStore: store,
+		resources:       table{},
+		waiter:          make(chan struct{}),
+	}
+}
+
+// Deactivate signals the store to close itself. Resources can not be accessed anymore,
+// but in progress resource updates are still active, until they are eventually ACKed.
+// The underlying persistent store will be finally closed once all pending updates
+// have been written to the persistent store.
+func (s *Store) Deactivate() {
+	s.resourcesMux.Lock()
+	defer s.resourcesMux.Unlock()
+	s.active = false
+}
+
+// Close deactivates the store and waits for all resources to be released.
+// It returns after the persistent store has been closed.
+func (s *Store) Close() {
+	s.Deactivate()
+	<-s.waiter
+}
+
+func (s *Store) shutdown() {
+	defer close(s.waiter)
+	s.persistentStore.Close()
+}
+
+// Access an unlocked resource. This creates a handle to a resource that may
+// not yet exist in the persistent registry.
+func (s *Store) Access(key ResourceKey) *Resource {
+	return newResource(s, key)
+}
+
+// Lock locks and returns the resource for a given key.
+func (s *Store) Lock(key ResourceKey) *Resource {
+	res := s.Access(key)
+	res.Lock()
+	return res
+}
+
+// TryLock locks and returns the resource for a given key.
+// The locked return value is set to false if TryLock failed, but the resource
+// itself is always returned.
```

Comment:
Why we would return a resource we cannot lock? I guess we would need to TryLock on the resource itself when need it?

Reply:
The code allows some read-only access to said resource. Although it is not locked, one could still query some of its internal status. The Lock/Unlock/TryLock methods on the store are just for convenience. Maybe I should remove them and always ask devs to call `Access` first.

Reply:
I would prefer to go through `Access` this effectively limit the possible bleeding of the api.

Reply:
I removed Lock/TryLock from the Store. Using `Access` is a must now.